syntax = "proto3";

package user_service;

option go_package = "grpc-vulnerable/proto";

// User service with vulnerable authorization
service UserService {
  // Authentication - works correctly
  rpc Login(LoginRequest) returns (LoginResponse);
  
  // Get user profile - requires authentication
  rpc GetUserProfile(GetUserProfileRequest) returns (UserProfile);
  
  // VULNERABLE: Admin-only operations without proper authorization
  rpc ListAllUsers(ListAllUsersRequest) returns (ListAllUsersResponse);
  rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse);
  rpc UpdateSystemSettings(UpdateSystemSettingsRequest) returns (UpdateSystemSettingsResponse);
  
  // VULNERABLE: Missing ownership check
  rpc UpdateUserSettings(UpdateUserSettingsRequest) returns (UpdateUserSettingsResponse);
  
  // VULNERABLE: Missing role check
  rpc ModerateContent(ModerateContentRequest) returns (ModerateContentResponse);
  
  // VULNERABLE: Information disclosure
  rpc GetUserDetails(GetUserDetailsRequest) returns (UserDetails);
}

// Request/Response messages
message LoginRequest {
  string username = 1;
  string password = 2;
}

message LoginResponse {
  string token = 1;
  UserProfile user = 2;
}

message GetUserProfileRequest {
  string token = 1;
}

message UserProfile {
  int32 user_id = 1;
  string username = 2;
  string email = 3;
  string role = 4;
}

message ListAllUsersRequest {
  string token = 1;
}

message ListAllUsersResponse {
  repeated UserProfile users = 1;
}

message DeleteUserRequest {
  string token = 1;
  int32 user_id = 2;
}

message DeleteUserResponse {
  bool success = 1;
  string message = 2;
}

message UpdateSystemSettingsRequest {
  string token = 1;
  string setting = 2;
  string value = 3;
}

message UpdateSystemSettingsResponse {
  bool success = 1;
  string message = 2;
}

message UpdateUserSettingsRequest {
  string token = 1;
  int32 target_user_id = 2;
  string setting = 3;
  string value = 4;
}

message UpdateUserSettingsResponse {
  bool success = 1;
  string message = 2;
}

message ModerateContentRequest {
  string token = 1;
  string content_id = 2;
  string action = 3;
}

message ModerateContentResponse {
  bool success = 1;
  string message = 2;
}

message GetUserDetailsRequest {
  string token = 1;
  int32 user_id = 2;
}

message UserDetails {
  int32 user_id = 1;
  string username = 2;
  string email = 3;
  string role = 4;
  string password_hash = 5;  // VULNERABLE: Should not be exposed
  string api_key = 6;        // VULNERABLE: Should not be exposed
  string created_at = 7;
}
